// theme-ui/dist/theme-ui.cjs.prod.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var core = require('@theme-ui/core');
var colorModes = require('@theme-ui/color-modes');
var mdx = require('@theme-ui/mdx');
var themeProvider = require('@theme-ui/theme-provider');
var components = require('@theme-ui/components');
var css = require('@theme-ui/css');

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function ownKeys(object, enumerableOnly) {
  var keys = Object.keys(object);

  if (Object.getOwnPropertySymbols) {
    var symbols = Object.getOwnPropertySymbols(object);

    if (enumerableOnly) {
      symbols = symbols.filter(function (sym) {
        return Object.getOwnPropertyDescriptor(object, sym).enumerable;
      });
    }

    keys.push.apply(keys, symbols);
  }

  return keys;
}

function _objectSpread2(target) {
  for (var i = 1; i < arguments.length; i++) {
    var source = arguments[i] != null ? arguments[i] : {};

    if (i % 2) {
      ownKeys(Object(source), true).forEach(function (key) {
        _defineProperty(target, key, source[key]);
      });
    } else if (Object.getOwnPropertyDescriptors) {
      Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
      ownKeys(Object(source)).forEach(function (key) {
        Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
      });
    }
  }

  return target;
}

var BaseStyles = function BaseStyles(props) {
  return jsx('div', _objectSpread2(_objectSpread2({}, props), {}, {
    sx: {
      fontFamily: 'body',
      lineHeight: 'body',
      fontWeight: 'body',
      variant: 'styles'
    }
  }));
};
var jsx = core.jsx;

Object.keys(components).forEach(function (k) {
  if (k !== 'default') Object.defineProperty(exports, k, {
    enumerable: true,
    get: function () {
      return components[k];
    }
  });
});
Object.defineProperty(exports, '__ThemeUIContext', {
  enumerable: true,
  get: function () {
    return core.__ThemeUIContext;
  }
});
Object.defineProperty(exports, 'createElement', {
  enumerable: true,
  get: function () {
    return core.createElement;
  }
});
Object.defineProperty(exports, 'merge', {
  enumerable: true,
  get: function () {
    return core.merge;
  }
});
Object.defineProperty(exports, 'useThemeUI', {
  enumerable: true,
  get: function () {
    return core.useThemeUI;
  }
});
Object.defineProperty(exports, 'InitializeColorMode', {
  enumerable: true,
  get: function () {
    return colorModes.InitializeColorMode;
  }
});
Object.defineProperty(exports, 'useColorMode', {
  enumerable: true,
  get: function () {
    return colorModes.useColorMode;
  }
});
Object.defineProperty(exports, 'Styled', {
  enumerable: true,
  get: function () {
    return mdx.Styled;
  }
});
Object.defineProperty(exports, 'Themed', {
  enumerable: true,
  get: function () {
    return mdx.Themed;
  }
});
Object.defineProperty(exports, 'components', {
  enumerable: true,
  get: function () {
    return mdx.components;
  }
});
Object.defineProperty(exports, 'ThemeProvider', {
  enumerable: true,
  get: function () {
    return themeProvider.ThemeProvider;
  }
});
Object.defineProperty(exports, 'css', {
  enumerable: true,
  get: function () {
    return css.css;
  }
});
Object.defineProperty(exports, 'get', {
  enumerable: true,
  get: function () {
    return css.get;
  }
});
exports.BaseStyles = BaseStyles;
exports.jsx = jsx;
